### ISSUE TITLE
The type constraint on Hash[Type] doesn't work

### ISSUE THREAD

**Issue by CIAvash** [rakudo/rakudo#2348]
## The Problem

`Hash[Int].new('a', 'b')` doesn't show a type check failure error.

## Expected Behavior

Show a Type Check Failure error:
"Type check failed in assignment to ; expected Int but got Str ("b")"

## Steps to Reproduce

```perl6
dd Hash[Int].new('a', 'b') # Hash[Int] element = (my Int % = :a("b")) 
```

## Environment
* Operating system: Arch Linux
* Compiler version (`perl6 -v`): 
This is Rakudo version 2018.09 built on MoarVM version 2018.09
implementing Perl 6.c.

**Comment by lizmat** [comment]
It appears that in case, it is calling the `STORE_AT_KEY` of a normal hash, rather than of the typed hash.

**Comment by lizmat** [comment]
It feels to me that `Hash.^parameterize` is not setting the correct `$!descriptor`.  But I'm at a loss as to how to **create** the correct descriptor.

**Comment by lizmat** [comment]
An illustration of the issue:

    my Int %h; use nqp; dd nqp::getattr(%h,Hash,q/$!descriptor/).of    # Int

    my %h := Hash[Int].new; use nqp; dd nqp::getattr(%h,Hash,q/$!descriptor/).of    # Mu

I tried adding a correct descriptor in `.STORE` when `:$initialize` is set, but that appears not to work :-(

**Comment by lizmat** [comment]
Fixed with 9acbf00041  and ea71041424 , tests added with https://github.com/perl6/roast/commit/28301d7c5d

Thanks for your report!
